servant/src/Servant/API/QueryParam.hs

```diff
@@ -38,6 +38,39 @@ data QueryParam' (mods :: [*]) (sym :: Symbol) (a :: *)
 data QueryParams (sym :: Symbol) (a :: *)
     deriving Typeable
 
+-- | Lookup the values associated with a collection of query string parameters
+-- and try to extract them as a value of type @a@. This is typically
+-- meant to query string parameters of the form
+-- @param1=val1&param2=val2@ and so on into a custom type represented by the form.
+--
+-- Note: Unlike with 'QueryParam', by default 'QueryParamForm' is parsed in a
+-- 'Lenient' way, because it's difficult to know if it should be parsed
+-- or not (when other 'QueryParam's are present). As a result, most users
+-- of 'QueryParamForm' are going to implement handlers that take a value
+-- of type (Maybe (Either Text a)). This also means that in a server implementation
+-- if there as a query string of any length (even just a "?"), we'll try to parse
```

Comment:
This seems confusing to me. If `QueryParamForm` is `Optional` we should **always** match that route then, independently whether there is `?` or not.

The difference between `http://example.com/` and `http://example.com/?` is too subtle.

Reply:
I'm not sure I understand the critique about matching routes because they both match: in the first case the handler will get a `Nothing` while in the second the Handler will get a `Just . Left a`. The routes do match, though?

This PR has always been based on the original PR which had been abandoned in an incomplete state, so I wouldn't take personally any criticism of it. In fact, I think there's a general criticism worth making that this implementation may cause some surprise among users. One thing I thought _may_ help is to define different, somehow more specific names for the `Optional` and `Required` variants. 

Another idea would be for the `Optional` case to return a `Maybe form` instead of a nested `Maybe (Either a form))`, in other words to give the user a `Just form` only in the `Just (Right form)` case and `Nothing` everywhere else. You would probably have to surrender the `Lenient`, `Strict` axis in that case and just say, "it's there or not there," but that may be confusing too and lead to buggy behavior if a query param is submitted misspelled or mistyped. Actually, this goes back to that conversation @alpmestan and I had above at some point in the hazy past...